ENH: random: Add the two-sided geometric distribution

andrewreed opened the pull request:
The new method

  two_sided_geometric(alpha, size=None)

of the class numpy.random.Generator implements the two-sided geometric
distribution, also known as the discrete Laplace distribution; see
  https://mathoverflow.net/questions/213221/
for an overview of how it differs from the "standard" geometric
distribution.

The user specifies the distribution's rate of decay with the `alpha`
parameter, where `alpha` is restricted to the interval [0,1).

andrewreed committed:
ENH: random: Add the two-sided geometric distribution

The new method

  two_sided_geometric(alpha, size=None)

of the class numpy.random.Generator implements the two-sided geometric
distribution, also known as the discrete Laplace distribution; see
  https://mathoverflow.net/questions/213221/
for an overview of how it differs from the "standard" geometric
distribution.

The user specifies the distribution's rate of decay with the `alpha`
parameter, where `alpha` is restricted to the interval [0,1).

bashtage commented on numpy/random/mtrand.pyx:
```diff
@@ -3481,6 +3482,99 @@ cdef class RandomState:
         # Match historical output type
         return int64_to_long(out)
 
+    def two_sided_geometric(self, alpha, size=None):
```
You must not change mtrand.pyx.  This frozen is effectively frozen, and will only receive changes necessary for it to compile in the future.

bashtage commented on numpy/random/src/distributions/distributions.c:
```diff
@@ -957,6 +957,39 @@ RAND_INT_TYPE random_geometric(bitgen_t *bitgen_state, double p) {
   }
 }
 
+RAND_INT_TYPE random_two_sided_geometric(bitgen_t *bitgen_state, double alpha) {
```
Should be int64_t.  No need for RAND_INT_TYPE since only available as part of Generator

bashtage commented on numpy/random/src/distributions/distributions.c:
```diff
@@ -957,6 +957,39 @@ RAND_INT_TYPE random_geometric(bitgen_t *bitgen_state, double p) {
+     * If the distribution does not return 0,
+     * then the result can be either + or -.
+     */
+    prob = next_double(bitgen_state);
```
Get the next uint32 and use the MSB (or some other bit that is not one of the 4 LSBs)

bashtage replied:
Something like 
```
sign = 2 * (int32_t)(next_uint32(bitgen_state) >> 31) - 1
```

bashtage commented on numpy/random/tests/test_random.py:
```diff
@@ -599,6 +599,14 @@ def test_geometric(self):
                             [5, 12]])
         assert_array_equal(actual, desired)
 
+    def test_two_sided_geometric(self):
```
I think these tests should go as well since you cannot change mtrand.pyx

bashtage reviewed (commented):
Biggest issue are the changes to mtrand.pyx which are not allowed under the NEP.  RandomState is frozen and gets no new features.

mattip commented:
Is this a commonly used distribution for typical statistical tests or experiments? My intuition would be that SciPy would be a better place for this.

andrewreed committed:
ENH: random: Add the two-sided geometric distribution

The new method

  two_sided_geometric(alpha, size=None)

of the class numpy.random.Generator implements the two-sided geometric
distribution, also known as the discrete Laplace distribution; see
  https://mathoverflow.net/questions/213221/
for an overview of how it differs from the "standard" geometric
distribution.

The user specifies the distribution's common ratio with the `alpha`
parameter, where `alpha` is restricted to the interval [0,1).

This commit removes the added function from mtrand.pyx, optimizes
the assignment for the 'sign' variable, and refines the function's
documentation.

andrewreed commented:
> Is this a commonly used distribution for typical statistical tests or experiments? My intuition would be that SciPy would be a better place for this.

I came across this distribution in my own research as the "two-sided distribution," and only noticed after I had started working on this PR that it can also be referred to as the "discrete Laplace distribution." I then saw that the discrete Laplace is already in SciPy, albeit with an alternate formulation of the PMF. IMHO, the PMF I found, with parameter 'alpha', I feel is a little more intuitive to use.

Given all of that, I concede that this PR may be of limited utility, especially within NumPy. Regardless, I appreciate the code review. This has been a good learning experience for me.

WarrenWeckesser commented:
@andrewreed, thanks for the contribution.  I think we're fairly conservative about expanding the distributions in `numpy.random`, and I suspect this one isn't used widely enough to warrant including in NumPy.    @rkern, what do you think?

andrewreed commented:
Before this PR is closed, I was wondering if I could get some additional insight:

I did some (totally informal) testing on two similarly resourced VMs and found that my algorithm inside NumPy is able to generate 10 million samples in ~0.3 secs, as opposed to SciPy's dlaplace which takes ~1.3 secs.  Furthermore, 100 million samples takes NumPy ~4 secs, whereas SciPy crashes with the message "Killed."

It also looks like the SciPy implementation of dlaplace relies on inverting the CDF, whereas my algorithm directly leverages the custom algorithm for geometric.

My two questions are:
1) Is the speed difference likely due to relying on inverting the CDF, or is there something inherently slower about sampling in SciPy?
2) Assuming there is a real speed-up with my implementation, is this the sort of change that would be worth trying to add to SciPy?

Again, thanks to all for your assistance.

rkern commented:
Yes, since no `dlaplace._rvs()` method is implemented, the default will draw a sample via inversion, which is indeed slow. It looks like this would be easy to implement `dlaplace._rvs()` efficiently in Python (draw from the corresponding geometric distribution and draw a fair coin toss to determine the sign). I'd implement this there.

rkern commented:
Generally, I would not add new distributions that are this straightforward to efficiently sample via composition of other existing distributions at the Python level. There are a number of such here already, but only because we needed them implemented in this API when it was first introduced so that it could drop-in replace the super-old Numeric and scipy (the long-gone `rv` module, not the `scipy.stats` distribution objects) PRNG libraries that we were trying to deprecate.

andrewreed commented:
> 
> 
> Yes, since no `dlaplace._rvs()` method is implemented, the default will draw a sample via inversion, which is indeed slow. It looks like this would be easy to implement `dlaplace._rvs()` efficiently in Python (draw from the corresponding geometric distribution and draw a fair coin toss to determine the sign). I'd implement this there.

Thanks!  On that note, I've been looking at SciPy and see the following under the geometric distribution:
```
def _rvs(self, p):
    return self._random_state.geometric(p, size=self._size)
```
Should I assume that, in this case, SciPy is actually using the NumPy implementation of geometric?

I ask because I think part of the efficiency in my algorithm is this initial check:
```
if (prob < (1.0 - alpha)/(1.0 + alpha)) {
  return 0;
}
```
This is required to calculate the probability of zero (the main difference between the standard geometric distribution), and also allows the function to return early for the samples that return 0 (i.e. no need to run the geometric function).

rkern commented:
Yes, `geom._rvs()` uses the `np.random.geometric()` implementation, which is what I recommend you use to build `dlaplace._rvs()`. I don't think the early-exit is going to be a major factor. Barring major errors in my quick implementation, it looks like a straightforward Python implementation has about the same speedup over the PMF inversion as your C implementation.

```

[~]
|7> import numpy as np

[~]
|8> from scipy import stats

[~]
|9> def two_sided_geometric_rvs(alpha, prng=np.random, size=None):
..>     prob = prng.random(size=size)
..>     sign = 2 * prng.randint(2, size=size) - 1
..>     x = sign * prng.geometric(1.0 - alpha)
..>     x[prob < (1-alpha) / (1+alpha)] = 0
..>     return x
..> 

[~]
|10> %timeit two_sided_geometric_rvs(0.5, size=10000000)
1 loop, best of 3: 211 ms per loop

[~]
|11> %timeit stats.dlaplace.rvs(0.5, size=10000000)
1 loop, best of 3: 1.1 s per loop
```

andrewreed commented:
Thanks, everyone!
